diff --git a/src/entry_point.rs b/src/entry_point.rs
index a24958c..e907e33 100644
--- a/src/entry_point.rs
+++ b/src/entry_point.rs
@@ -348,7 +348,7 @@ pub unsafe extern "C" fn rust_start(app_start: usize, stacktop: usize, app_heap_
     // Heap size is set using `elf2tab` with `--app-heap` option, which is
     // currently at 1024. If you change the `elf2tab` heap size, make sure to
     // make the corresponding change here.
-    const HEAP_SIZE: usize = 1024;
+    const HEAP_SIZE: usize = 90000;
 
     // we could have also bss_end for app_heap_start
     let app_heap_start = app_heap_break;
