diff --git a/Cargo.toml b/Cargo.toml
index 83c48558e..75b1abacc 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -12,7 +12,11 @@ members = [
     "boards/imix",
     "boards/msp_exp432p401r",
     "boards/nordic/nrf52840dk",
+    "boards/nordic/nrf52840dk_opensk",
     "boards/nordic/nrf52840_dongle",
+    "boards/nordic/nrf52840_dongle_opensk",
+    "boards/nordic/nrf52840_dongle_dfu",
+    "boards/nordic/nrf52840_mdk_dfu",
     "boards/nordic/nrf52dk",
     "boards/nucleo_f429zi",
     "boards/nucleo_f446re",
