diff --git a/Cargo.toml b/Cargo.toml
index 83c4855..cfd63b1 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -13,6 +13,8 @@ members = [
     "boards/msp_exp432p401r",
     "boards/nordic/nrf52840dk",
     "boards/nordic/nrf52840_dongle",
+    "boards/nordic/nrf52840_dongle_dfu",
+    "boards/nordic/nrf52840_mdk_dfu",
     "boards/nordic/nrf52dk",
     "boards/nucleo_f429zi",
     "boards/nucleo_f446re",
