diff --git a/boards/nordic/nrf52840_dongle/layout.ld b/boards/nordic/nrf52840_dongle/layout.ld
index 657b0d26..f86b2321 100644
--- a/boards/nordic/nrf52840_dongle/layout.ld
+++ b/boards/nordic/nrf52840_dongle/layout.ld
@@ -1,6 +1,6 @@
 MEMORY
 {
-  rom (rx)  : ORIGIN = 0x00000000, LENGTH = 128K
+  rom (rx)  : ORIGIN = 0x00000000, LENGTH = 192K
   prog (rx) : ORIGIN = 0x00030000, LENGTH = 832K
   ram (rwx) : ORIGIN = 0x20000000, LENGTH = 256K
 }
diff --git a/boards/nordic/nrf52840dk/layout.ld b/boards/nordic/nrf52840dk/layout.ld
index 657b0d26..f86b2321 100644
--- a/boards/nordic/nrf52840dk/layout.ld
+++ b/boards/nordic/nrf52840dk/layout.ld
@@ -1,6 +1,6 @@
 MEMORY
 {
-  rom (rx)  : ORIGIN = 0x00000000, LENGTH = 128K
+  rom (rx)  : ORIGIN = 0x00000000, LENGTH = 192K
   prog (rx) : ORIGIN = 0x00030000, LENGTH = 832K
   ram (rwx) : ORIGIN = 0x20000000, LENGTH = 256K
 }
