diff --git a/kernel/src/sched.rs b/kernel/src/sched.rs
index ed3ae8260..f109960d3 100644
--- a/kernel/src/sched.rs
+++ b/kernel/src/sched.rs
@@ -632,6 +632,7 @@ impl Kernel {
                     let context_switch_reason = process.switch_to();
                     scheduler_timer.disarm();
                     chip.mpu().disable_app_mpu();
+                    chip.mpu().clear_mpu();
 
                     // Now the process has returned back to the kernel. Check
                     // why and handle the process as appropriate.
