diff --git a/arch/cortex-m/src/syscall.rs b/arch/cortex-m/src/syscall.rs
index c78b1c9fb..2769d0138 100644
--- a/arch/cortex-m/src/syscall.rs
+++ b/arch/cortex-m/src/syscall.rs
@@ -65,7 +65,7 @@ impl kernel::syscall::UserspaceKernelBoundary for SysCall {
         //
         // The 1.x Tock kernel allocates at least 3 kB to processes, and we need
         // to ensure that happens as userspace may expect it.
-        3 * 1024
+        70 * 1024
 
         // TOCK 2.0
         //
