diff --git a/capsules/src/usb/usb_ctap.rs b/capsules/src/usb/usb_ctap.rs
index e8f1a87a4..2c1ecf934 100644
--- a/capsules/src/usb/usb_ctap.rs
+++ b/capsules/src/usb/usb_ctap.rs
@@ -57,7 +57,7 @@ impl<'a, 'b, C: hil::usb::UsbController<'a>> CtapUsbSyscallDriver<'a, 'b, C> {
                 app.waiting = false;
                 // Signal to the app that a packet is ready.
                 app.callback
-                    .map(|mut cb| cb.schedule(CTAP_CALLBACK_RECEIVED, 0, 0));
+                    .map(|mut cb| cb.schedule(CTAP_CALLBACK_RECEIVED, 1, 0));
             }
         }
     }
